libraries/tock-register-interface/README.md

```diff
@@ -3,130 +3,107 @@
-convention using a field named `_reservedN`), but without a type. The macro will
-then automatically take care of calculating the gap size and inserting a
-suitable filler struct. The end of the struct is marked with its size and the
-`@END` keyword, effectively pointing to the offset immediately past the list of
-registers.
+## Defining peripherals
 
-```rust
-use tock_registers::registers::{ReadOnly, ReadWrite, WriteOnly};
+An MMIO peripheral is defined using the `peripheral!` macro:
 
-register_structs! {
+```rust
+tock_registers::peripheral! {
     Registers {
-        // Control register: read-write
-        // The 'Control' parameter constrains this register to only use fields from
-        // a certain group (defined below in the bitfields section).
-        (0x000 => cr: ReadWrite<u8, Control::Register>),
-
-        // Status register: read-only
-        (0x001 => s: ReadOnly<u8, Status::Register>),
-
-        // Registers can be bytes, halfwords, or words:
-        // Note that the second type parameter can be omitted, meaning that there
-        // are no bitfields defined for these registers.
-        (0x002 => byte0: ReadWrite<u8>),
-        (0x003 => byte1: ReadWrite<u8>),
-        (0x004 => short: ReadWrite<u16>),
+        // Control register: read-write. `u8` specifies the data type of the
+        // memory-mapped register, and the `Control` parameter specifies how
+        // bitfields within the register should be accessed.
+        0x000 => cr: u8(Control::Register) { Read, Write },
+
+        // Status register: read-only. Registers can have documentation
+        // comments, like:
+        /// Status register
+        0x001 => s: u8(Status::Register) { Read },
+
+        // Registers can have different sizes. The bitfield parameter is
+        // optional:
+        0x002 => byte0: u8 { Read, Write },
+        0x003 => byte1: u8 { Read, Write },
+        0x004 => short: u16 { Read, Write },
 
         // Empty space between registers must be marked with a padding field,
-        // declared as follows. The length of this padding is automatically
-        // computed by the macro.
-        (0x006 => _reserved),
-        (0x008 => word: ReadWrite<u32>),
-
-        // The type for a register can be anything. Conveniently, you can use an
-        // array when there are a bunch of similar registers.
-        (0x00C => array: [ReadWrite<u32>; 4]),
-        (0x01C => ... ),
-
-        // Etc.
-
-        // The end of the struct is marked as follows.
-        (0x100 => @END),
+        // specified by naming the field _ and omitting the rest of the
+        // specification. The length of the padding field will be inferred. In
+        // practice, padding fields remove the check that two successive
+        // registers are adjacent in the address space.
+        0x006 => _,
+
+        // If you have several adjacent registers, you can specify an array
+        // register:
+        0x00C => gpio_pins: [u32; 4] { Read, Write },
```

Comment:
What is the syntax for an array of bitfield registers?

Reply:
Oh, I just answered this in a different comment chain: https://github.com/tock/tock/pull/4001#discussion_r1617915028